parser.js

```diff
@@ -25,15 +25,41 @@ const both = R.curry((p1, p2) =>
   )
 );
 
-// many :: Parser a -> Parser [a]
-const many = R.curry(parser => string => {
-  parser(string).chain();
-});
-
 const seq = R.curry((...parsers) => parsers.reduce(both));
 
 const result = R.curry((result, remainder) => Just({ result, remainder }));
 
+// chainP :: Parser a -> (a -> Parser b) -> Parser b
+const chainP = R.curry((parser, step, string) =>
+  parser(string).chain(({result, remainder}) =>
+    step(result)(remainder)
+  )
+);
+
+// parse 0 or more occurrences (always succeeds, e.g. with [] if no matches)
+// many0 :: Parser a -> Parser [a]
+const many0 = R.curry(parser => string =>
+  chainP(
+    parser,
+    x => chainP(
+      many0(parser),
+      xs => result([x, ...xs])
```

Comment:
Note that this spread syntax is not very performant given the recursive nature of the function. One solution would be to use a DList and then `map` the DList to an array at the very end. Of course, the entire parser combinator library itself is not necessarily performant; to make it performant requires a more complex implementation. But I presume you are writing this for fun / self-education!